spec/abstract/IpcCancel_A.thy

```diff
@@ -316,28 +316,47 @@ where
     BlockedOnNotification r \<Rightarrow> Some r
   | _ \<Rightarrow> None"
 
-text \<open>
-  Sort a list of TCB refs by priority, otherwise keeping order stable.
-  In the implementation, there will only ever be one element out of order
-  or be inserted. Here, we just require a sorted result.
-\<close>
+fun
+  tcb_ep_find_index :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> nat \<Rightarrow> (nat, 'z::state_ext) s_monad"
+where
+  "tcb_ep_find_index tptr qs curindex = do
+     prio \<leftarrow> thread_get tcb_priority tptr;
+     curprio \<leftarrow> thread_get tcb_priority (qs ! curindex);
+     if prio > curprio then
+       if curindex = 0 then return 0
+       else tcb_ep_find_index tptr qs (curindex - 1)
+     else return (curindex + 1)
+   od"
+
+declare tcb_ep_find_index.simps[simp del]
+
 definition
-  sort_queue :: "obj_ref list \<Rightarrow> (obj_ref list, 'z::state_ext) s_monad"
+  tcb_ep_dequeue :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> (obj_ref list, 'z::state_ext) s_monad"
 where
-  "sort_queue qs = do
-     prios \<leftarrow> mapM (thread_get tcb_priority) qs;
-     return $ map snd $ sort_key (\<lambda>x. 255 - (fst x)) (zip prios qs) \<comment> \<open>@{text \<open>0 \<le> priority < 256\<close>}\<close>
+  "tcb_ep_dequeue tptr qs = do
+     index \<leftarrow> return $ the $ findIndex (\<lambda>x. x = tptr) qs;
+     return $ take index qs @ drop (index + 1) qs
    od"
 
+definition
+  tcb_ep_append :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> (obj_ref list, 'z::state_ext) s_monad"
```

Comment:
We should change the name of this function since it's no longer doing an append. Maybe `tcb_ep_insert`? The same change in the C and Haskell would also help.